cilium/CFP-12345-high-scale-ipcache.md

```diff
@@ -0,0 +1,116 @@
+2. The packet is encapsulated with "pod1 -> pod2" as the outer IP header and
+   the tunnel ID is set to pod1's security identity. See section "Pod-to-pod
+   encapsulation" below for details.
+3. Node2 receives the encapsulated packet, retrieves the source security
+   identity and decapsulates.
+4. The decapsulated packet is sent to pod2's ingress policy enforcement via
+   tail call and the ingress policies are enforced using the retrieved security
+   identity.
+
+### Changes to the IPcache
+
+The goal is to reduce the size of the ipcache and in particular for the ipcache
+to not grow linearly with the number of pods in the cluster.
+
+We will keep in the ipcache:
+- Special identities for entities (host, remote nodes, etc.).
+- Identities for well-known identities. This is mostly to get DNS pod
+  identities. If we want to later use FQDN policies on egress, we will need to
+  resolve domain names via the DNS pods. We can't use the FQDN policies to
+  allow connections to the DNS pods obviously so we'll have to rely on
+  toEndpoints for that.
+
+### Pod-to-pod Encapsulation
+
+In high-scale IPcache mode, the encapsulation for pod-to-pod traffic is only
+required to carry the source identity over to the destination node. However,
+since Cilium doesn't know how to map pods to their node IP addresses (since
+we don't have the ipcache), we can't use our usual encapsulation with node IPs.
+
+Instead, for the inner and outer headers, we will use the same IP addresses,
+the pod IP addresses. Setting the outer destination IP address is easy, we can
+use `bpf_skb_set_tunnel_key` as we do today. Setting the outer source IP
+address is only supported in Linux v5.19
+([26101f5ab6bd](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=26101f5ab6bd)).
+
+It would also have been possible to use the node IP address as the outer source
+IP. That would however result in an assymetric connection on the physical
+network. Forward packets would have IPs IP_NodeA -> IP_PodB and replies
+IP_NodeB -> IP_PodA. That may cause issues with middleboxes and observability
+tooling on the physical network.
+
+On ingress, we shouldn't need to do anything new to decapsulate the traffic. We
+already retrieve the identity in bpf_overlay and don't use the outer IP
+addresses anyway.
+
+### World CIDR CRD and Map
+
+To know what traffic needs to be encapsulated, we can't rely on the ipcache or
+the tunnel map (we don't have a clear podCIDR->node mapping for all IPAMs). We
+will need a new CRD for the users to provide the CIDRs of destination that
+don't need to be encapsulated. Everything else will be encapsulated.
+
+This new CRD is CiliumWorldCIDRs. It will provide a list of destination CIDRs
+for which no encapsulation is necessary, typically destination outside the
+cluster. This list of CIDRs will be plumbed into an LPM BPF map for the
+datapath to check if encapsulation is needed.
```

Comment:
I have a mild concern that we have multiple resources starting to vie for a very similar semantic in this space, and this could do with a bit more of a holistic view.

Specifically:
1. [IPMasqAgent config support](https://github.com/cilium/cilium/pull/11148) determines a set of prefixes "outside the cluster" to perform masquerade upon.
1. [CiliumCIDRGroup](https://github.com/cilium/cilium/pull/24638) allows a set of IPs to be configured in a resource as a "group of CIDRs outside the cluster"
1. This one is proposing a "CiliumWorldCIDRs" resource

(1) is the oldest, but AFAICT it operates on a ConfigMap resource, following a specification from [another k8s project](https://github.com/kubernetes-sigs/ip-masq-agent)
(2) was recently introduced, but it is a generic "CIDR group resource" that allows other resources to reference it. Notably, while it defines that the CIDRs inside are "outside the cluster", it doesn't explicitly define semantics for encap/no encap.

Given that (1) must retain compatibility with a pre-specified resource, and (2) provides "Set" semantics, I wonder about the form of this "world CIDRs" and how generic we may want it to be. Ideally we could group a set of similar or related datapath configurations together and define k8s-level semantics across something like the existing CiliumCIDRGroup resources (similar to how `cidrGroupRef` works in CNP) in order to allow set management to be handled in one resource, then features managed across that set in another resource, and then we need exactly these two resources.

If we don't have a good sense for the future-proofing of a more generic resource for this purpose, the danger would just be that we create this resource, then in another release or two we want to define yet another resource that applies some feature to some range of prefixes. Then we'll be in the position of asking whether we need to deprecate the resource proposed here, or change its semantics, etc. Not the end of the world by any means but if we do see a more generic way to represent this configuration today, it could save us a bit of pain later on.

Reply:
One possible shorter term approach would be that this CFP just introduces a new flag something like `--world-cidrs=cidrGroupRef:WorldCIDRs`, then this is simply a reference to the CiliumCIDRGroup resource that has the name `WorldCIDRs`. Then we know that the CiliumCIDRGroup set membership can be independently managed, we have just that one resource to subscribe to in k8s, and it's not quite as much overhead to put the feature together as creating a whole additional CRD for this specific use case.

Reply:
I'm not the best person to answer this. We would need someone from the agent to look into it, to be able to consider the pros and cons of managing those CRDs. One thought on that last proposal is that I would expect users to sometimes want to have several CiliumWorldCIDR CRs, with different group of external entities (syntactic sugar, no datapath impact).